README.rst
Python Client for Google BigQuery Reservation API
=================================================

|stable| |pypi| |versions|

`Google BigQuery Reservation API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-bigquery-reservation.svg
   :target: https://pypi.org/project/google-cloud-bigquery-reservation/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-bigquery-reservation.svg
   :target: https://pypi.org/project/google-cloud-bigquery-reservation/
.. _Google BigQuery Reservation API: https://cloud.google.com/bigquery/docs/reference/reservations
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/bigqueryreservation/latest
.. _Product Documentation:  https://cloud.google.com/bigquery/docs/reference/reservations

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Google BigQuery Reservation API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Google BigQuery Reservation API.:  https://cloud.google.com/bigquery/docs/reference/reservations
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-bigquery-reservation


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-bigquery-reservation

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Google BigQuery Reservation API
   to see other available methods on the client.
-  Read the `Google BigQuery Reservation API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Google BigQuery Reservation API Product documentation:  https://cloud.google.com/bigquery/docs/reference/reservations
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [1.11.3](https://github.com/googleapis/python-bigquery-reservation/compare/v1.11.2...v1.11.3) (2023-09-11)


### Documentation

* Minor formatting ([#377](https://github.com/googleapis/python-bigquery-reservation/issues/377)) ([1fd7e69](https://github.com/googleapis/python-bigquery-reservation/commit/1fd7e69a0eeb98de9fd3092247eb9d06147cbad6))

## [1.11.2](https://github.com/googleapis/python-bigquery-reservation/compare/v1.11.1...v1.11.2) (2023-07-05)


### Bug Fixes

* Add async context manager return types ([#362](https://github.com/googleapis/python-bigquery-reservation/issues/362)) ([98df6ad](https://github.com/googleapis/python-bigquery-reservation/commit/98df6adf73f2a1276e34977c06882a87558edf58))

## [1.11.1](https://github.com/googleapis/python-bigquery-reservation/compare/v1.11.0...v1.11.1) (2023-03-28)


### Documentation

* Minor clarifications ([#352](https://github.com/googleapis/python-bigquery-reservation/issues/352)) ([c9ce9b8](https://github.com/googleapis/python-bigquery-reservation/commit/c9ce9b85b8315f5d676fecdc0c95fd4f01363e9a))

## [1.11.0](https://github.com/googleapis/python-bigquery-reservation/compare/v1.10.0...v1.11.0) (2023-03-23)


### Features

* Add edition/autoscale related fields ([#348](https://github.com/googleapis/python-bigquery-reservation/issues/348)) ([e94a53b](https://github.com/googleapis/python-bigquery-reservation/commit/e94a53bd510f51facc4a0fd591e5c4a981028f90))


### Documentation

* Fix formatting of request arg in docstring ([#350](https://github.com/googleapis/python-bigquery-reservation/issues/350)) ([ecce362](https://github.com/googleapis/python-bigquery-reservation/commit/ecce36210c6a8fdce862b9ff16d49451180bea9a))

## [1.10.0](https://github.com/googleapis/python-bigquery-reservation/compare/v1.9.1...v1.10.0) (2023-02-16)


### Features

* Enable "rest" transport in Python for services supporting numeric enums ([#339](https://github.com/googleapis/python-bigquery-reservation/issues/339)) ([50de8ca](https://github.com/googleapis/python-bigquery-reservation/commit/50de8ca394e61e0fd97b9264c660d4017c241ece))

## [1.9.1](https://github.com/googleapis/python-bigquery-reservation/compare/v1.9.0...v1.9.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([3912795](https://github.com/googleapis/python-bigquery-reservation/commit/3912795237f099f8cae2dea647f6e599aa3c4a1b))


### Documentation

* Add documentation for enums ([3912795](https://github.com/googleapis/python-bigquery-reservation/commit/3912795237f099f8cae2dea647f6e599aa3c4a1b))

## [1.9.0](https://github.com/googleapis/python-bigquery-reservation/compare/v1.8.0...v1.9.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#327](https://github.com/googleapis/python-bigquery-reservation/issues/327)) ([c86b3b8](https://github.com/googleapis/python-bigquery-reservation/commit/c86b3b8da191595dcbb30b4a55136c30376bad72))

## [1.8.0](https://github.com/googleapis/python-bigquery-reservation/compare/v1.7.3...v1.8.0) (2022-12-15)


### Features

* Add support for `google.cloud.bigquery_reservation.__version__` ([e82b8e5](https://github.com/googleapis/python-bigquery-reservation/commit/e82b8e50be61e2885ac8f5b21f650e54f8d6d604))
* Add typing to proto.Message based class attributes ([e82b8e5](https://github.com/googleapis/python-bigquery-reservation/commit/e82b8e50be61e2885ac8f5b21f650e54f8d6d604))


### Bug Fixes

* Add dict typing for client_options ([e82b8e5](https://github.com/googleapis/python-bigquery-reservation/commit/e82b8e50be61e2885ac8f5b21f650e54f8d6d604))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([c9ff564](https://github.com/googleapis/python-bigquery-reservation/commit/c9ff564b070c488b96af8574df442334173996cf))
* Drop usage of pkg_resources ([c9ff564](https://github.com/googleapis/python-bigquery-reservation/commit/c9ff564b070c488b96af8574df442334173996cf))
* Fix timeout default values ([c9ff564](https://github.com/googleapis/python-bigquery-reservation/commit/c9ff564b070c488b96af8574df442334173996cf))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([e82b8e5](https://github.com/googleapis/python-bigquery-reservation/commit/e82b8e50be61e2885ac8f5b21f650e54f8d6d604))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([c9ff564](https://github.com/googleapis/python-bigquery-reservation/commit/c9ff564b070c488b96af8574df442334173996cf))

## [1.7.3](https://github.com/googleapis/python-bigquery-reservation/compare/v1.7.2...v1.7.3) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#313](https://github.com/googleapis/python-bigquery-reservation/issues/313)) ([e616e6b](https://github.com/googleapis/python-bigquery-reservation/commit/e616e6bc2be88305ba5088e37647bc934b28299e))

## [1.7.2](https://github.com/googleapis/python-bigquery-reservation/compare/v1.7.1...v1.7.2) (2022-10-03)


### Bug Fixes

* **deps:** Require protobuf >= 3.20.2 ([#310](https://github.com/googleapis/python-bigquery-reservation/issues/310)) ([9c5c59a](https://github.com/googleapis/python-bigquery-reservation/commit/9c5c59a6f8b164f4d323677be0de537616f2f664))

## [1.7.1](https://github.com/googleapis/python-bigquery-reservation/compare/v1.7.0...v1.7.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#292](https://github.com/googleapis/python-bigquery-reservation/issues/292)) ([112a726](https://github.com/googleapis/python-bigquery-reservation/commit/112a726680d4426bdbc037631c887915bb8b2e35))
* **deps:** require proto-plus >= 1.22.0 ([112a726](https://github.com/googleapis/python-bigquery-reservation/commit/112a726680d4426bdbc037631c887915bb8b2e35))

## [1.7.0](https://github.com/googleapis/python-bigquery-reservation/compare/v1.6.3...v1.7.0) (2022-07-16)


### Features

* add audience parameter ([8b25cff](https://github.com/googleapis/python-bigquery-reservation/commit/8b25cff82fab1ae6c038238df5d65318396c0782))


### Bug Fixes


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/bigquery_reservation_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
